Api controller bb

benbutler08 opened the pull request:
Basic crud controller with strong parameters, added admins as a resource.

benbutler08 committed:
CRUD

make a basic CRUD controller for the admins model.

benbutler08 committed:
strong parameters

Add private methd admin_params to use strong parameters

johnkeith reviewed (changes requested):
wheres the specs!

can you install rspec and add some?

johnkeith commented on app/controllers/api/v1/admins_controller.rb:
```diff
@@ -0,0 +1,48 @@
+class AdminsController < ApplicationController
```
also, we need to change the name:

since we are nested underneath /api Rails is going to need to see that in the class name.

```Api::V1::AdminsController```

johnkeith commented on app/controllers/api/v1/admins_controller.rb:
```diff
@@ -0,0 +1,48 @@
+class AdminsController < ApplicationController
+  def new
```
since this is an API, we will not need the #new...you'll see later on

johnkeith commented on app/controllers/api/v1/admins_controller.rb:
```diff
@@ -0,0 +1,48 @@
+  def create
+    @admin = Admin.new params(admin_params)
+    if @admin.save
+      redirect_to admins_path
```
let's render JSON! We are creating a JSON Api here to be used by our react app:

```render json: @admin```

so change all the renders and redirects to return JSON

johnkeith commented on app/controllers/api/v1/admins_controller.rb:
```diff
@@ -0,0 +1,48 @@
+    end
+  end
+
+  def edit
```
not needed

johnkeith commented on app/controllers/api/v1/admins_controller.rb:
```diff
@@ -0,0 +1,48 @@
+  def destroy
+    @admin = Admin.find params[:id]
+    @admin.destroy
+    redirect_to admins_path
```
here we typically render head :no_content on a destroy. Just a common practice with a JSON API.

benbutler08 committed:
name

Fix name to include Api and V1

benbutler08 committed:
no new or edit for api

Remove edit and new actions

benbutler08 committed:
set_admin

Add a private method as a before action to set @admin, finding it by id

benbutler08 committed:
api-ify returns

Change actions to return json or head: no_content (update and destroy)

benbutler08 committed:
render json

replace json_response with render json:

benbutler08 committed:
Rspec first spec

Add RSpec, create a controllers folder with a spec for admins controller

johnkeith commented on app/controllers/api/v1/admins_controller.rb:
```diff
@@ -45,4 +32,8 @@ def admin_params
   end
 
+  def set_admin
+    @admin = Admin.find(params[:id])
```
Nice!

benbutler08 committed:
nest resource in namespaces

Move the admins resource into the api namespaces, which should correct the routes. Add status to successful and unsuccessful update and destroy.

benbutler08 committed:
Passing spec

Change the controller spec to be a request. Change the routing to be /api/v1/admins. Is passing.

benbutler08 committed:
factory_bot

Add factory_bot, enable shorter syntax with FactoryBot::Syntax::Methods

benbutler08 committed:
cleaner spec

Change the admin_controller spec around a bit. Is now failing because of validations, admin needs account.

benbutler08 committed:
account factory

add an account factory and an association in the admins factory. Change factory names to singular.

johnkeith commented on spec/factories/account.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :account do
+    name 'Test Account'
+    id '1'
```
it should auto increment the ID I believe I think you can remove that

benbutler08 committed:
remove explicit account ids

ids should auto-increment

benbutler08 committed:
build not create

Change create :admin to build, remove ! from let. Add JSON.parse back into the expect param. Passing.

benbutler08 committed:
add validation test

test that an admin must have a name to be saved

benbutler08 committed:
name validation test

Add a valudation test for admin names

benbutler08 committed:
Add name presence validation

Add in a simple validation for name presence, validation test is now passing.

benbutler08 committed:
show action

create a test for the show action

benbutler08 committed:
refocus on index

Get rid of show and validation tests. index test now returns json, doesn't match yet.

benbutler08 committed:
response expectations fixed

Change the response expectations. passing.

benbutler08 committed:
change get index test

make sure it is returning all the admins

benbutler08 committed:
change index comparison

Change factoryBot call to create_list, check for length of the list created. passing.

benbutler08 committed:
Show action test

create an admin with an id of 1, request it from show. passing.

benbutler08 committed:
post #create test

Add a test for Post that creates an admin with a valid account_id.

benbutler08 committed:
randomize id created

Change id created for account to be a random number less than 1000.

benbutler08 committed:
test for unsuccessful post

change the random id generated to create an admin with the account_id of the first account. Add a test for when an admin is not created.

benbutler08 committed:
dry up the spec

collect Factories to the top, set before hooks

benbutler08 committed:
add faker and seeds

Seed the database with faked records

benbutler08 committed:
protect CSRF for apis

add protect_from_forgery with null_sessions to stop authenticity token errors with create

benbutler08 committed:
update test

Change test data to create a new model object before each test. Add update description.

benbutler08 committed:
#create change

Change the create action, add require(:admin) to strong params

benbutler08 committed:
Revert "#create change"

This reverts commit b9587329edfeebe1ba92d07d8560f0efee34d025.

benbutler08 committed:
change create

Change the create action, add require(:admin) to strong params

benbutler08 committed:
passing create test

change the create test to take attributes from a factory_bot object build. Passing.

benbutler08 committed:
Update test

Update test, getting NoMethodError, it doesn't seem to be finding the admin to update

johnkeith committed:
happy specs

benbutler08 committed:
destroy test

Add test for destroy action. Passing.
